### ISSUE TITLE
onFieldSubmitted not triggered in windows

### ISSUE THREAD

**Issue by KaushikGupta007** [flutter/flutter#150338]
### Steps to reproduce

1. Create a simple flutter app with TextFormField on Home Page
2. Run app on windows flutter
3. Click on field
4. Press enter
5. onFieldSubmitted is not triggered

### Expected results

onFieldSubmitted should be triggered

### Actual results

onFieldSubmitted is not triggered

### Code sample

<details open><summary>Code sample</summary>

```dart
Future<void> main() async {
  runApp(MaterialApp(
    initialRoute: "/",
    routes: {
      "/": (context) => const Home(),
    },
  ));
}

class Home extends StatefulWidget {
  const Home({super.key});

  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends State<Home> {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Flutter v3.22.2"),
      ),
      body: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Column(
          children: [
            TextFormField(
              decoration: const InputDecoration(hintText: "First Field", labelText: "First Field"),
              readOnly: true,
              onFieldSubmitted: (val){
                var snackBar = const SnackBar(content: Text('Hello World'));
                ScaffoldMessenger.of(context).showSnackBar(snackBar);
              },
            )
          ],
        ),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

**Expected Result in Flutter Web**

https://github.com/flutter/flutter/assets/129219828/7750ee0d-1daf-4845-bef9-2017994921b9

**Actual Results in Flutter Windows**

https://github.com/flutter/flutter/assets/129219828/abd7fdb0-2231-49c9-b72d-1332b59a041d

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[√] Flutter (Channel stable, 3.22.2, on Microsoft Windows [Version 10.0.22621.3593], locale en-IN)
    • Flutter version 3.22.2 on channel stable at C:\Users\Admin\fvm\versions\3.22.2
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 761747bfc5 (12 days ago), 2024-06-05 22:15:13 +0200
    • Engine revision edd8546116
    • Dart version 3.4.3
    • DevTools version 2.34.3

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at C:\Users\Admin\AppData\Local\Android\sdk
    • Platform android-34, build-tools 33.0.1
    • Java binary at: C:\Program Files\Android\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-9586694)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.9.1)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.9.34616.47
    • Windows 10 SDK version 10.0.22621.0

[√] Android Studio (version 2022.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-9586694)

[√] VS Code (version 1.89.1)
    • VS Code at C:\Users\Admin\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.90.0

[√] Connected device (5 available)
    • CPH2467 (mobile)                   • 1a715d74      • android-arm64  • Android 14 (API 34)
    • Android SDK built for x86 (mobile) • emulator-5554 • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)                  • windows       • windows-x64    • Microsoft Windows [Version 10.0.22621.3593]
    • Chrome (web)                       • chrome        • web-javascript • Google Chrome 126.0.6478.61
    • Edge (web)                         • edge          • web-javascript • Microsoft Edge 125.0.2535.85

[√] Network resources
    • All expected network resources are available.

• No issues found!

```

</details>

**Comment by huycozy** [comment]
Hi @KaushikGupta007 
I think this is intentional as you are also putting `readOnly: true` as well. I'm more inclined to think this is an issue with Web instead. 

Can you share a use case of this when a field is read-only but users still press `Enter` key? This behavior seems unusual.

**Comment by KaushikGupta007** [comment]
Hi @huycozy 
So basically i am developing an Accounting Software. While creating an Invoice i have _Amount_ field value for which is calculated using _Quantity times Rate_. I don't want this value to change, so it is readOnly field. Now, i also want User to sort of confirm Final Amount before adding it to the invoice, so cursor should go to the field and on Pressing Enter add data to Pluto Table.
Please refer video for clarification.

https://github.com/flutter/flutter/assets/129219828/c6fbbe82-3baa-4a84-98f8-8a358187dfca

**Comment by KaushikGupta007** [comment]
Since read only fields are capable of being focused, so `onFieldSubmitted` should be triggered on platforms that use physical keyboard like web, windows or macOS

**Comment by huycozy** [comment]
Thanks for your update. I'm seeing the same result as reported on macOS and Linux app targets as well. 

<details>
<summary>flutter doctor -v (stable & master)</summary>

```bash
[✓] Flutter (Channel stable, 3.22.2, on Ubuntu 22.04.3 LTS 6.5.0-26-generic, locale en_US.UTF-8)
    • Flutter version 3.22.2 on channel stable at /home/huynq/Documents/Working/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 761747bfc5 (5 days ago), 2024-06-05 22:15:13 +0200
    • Engine revision edd8546116
    • Dart version 3.4.3
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0-rc3)
    • Android SDK at /home/huynq/Android/Sdk/
    • Platform android-33, build-tools 34.0.0-rc3
    • Java binary at: /snap/android-studio/current/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 14.0.0-1ubuntu1.1
    • cmake version 3.22.1
    • ninja version 1.10.1
    • pkg-config version 0.29.2

[✓] Android Studio (version 2021.1)
    • Android Studio at /home/huynq/Documents/android-studio
    • Flutter plugin version 67.0.1
    • Dart plugin version 211.7817
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)

[!] Android Studio (version unknown)
    • Android Studio at /snap/android-studio/current
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • android-studio-dir = /snap/android-studio/current
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] VS Code (version 1.84.2)
    • VS Code at /usr/share/code
    • Flutter extension version 3.90.0

[✓] VS Code (version 1.89.1)
    • VS Code at /snap/code/current/usr/share/code
    • Flutter extension version 3.90.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 22.04.3 LTS 6.5.0-26-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 120.0.6099.199

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

```bash
[!] Flutter (Channel master, 3.23.0-13.0.pre.261, on Ubuntu 22.04.3 LTS 6.5.0-35-generic, locale en_US.UTF-8)
    • Flutter version 3.23.0-13.0.pre.261 on channel master at /home/huynq/Documents/Working/flutter_master
    ! Warning: `flutter` on your path resolves to /home/huynq/Documents/Working/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /home/huynq/Documents/Working/flutter_master. Consider adding /home/huynq/Documents/Working/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /home/huynq/Documents/Working/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /home/huynq/Documents/Working/flutter_master. Consider adding /home/huynq/Documents/Working/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fb067b9948 (2 hours ago), 2024-06-17 21:41:41 -0400
    • Engine revision 1c4e5e230e
    • Dart version 3.5.0 (build 3.5.0-271.0.dev)
    • DevTools version 2.36.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0-rc3)
    • Android SDK at /home/huynq/Android/Sdk/
    • Platform android-33, build-tools 34.0.0-rc3
    • Java binary at: /snap/android-studio/current/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 14.0.0-1ubuntu1.1
    • cmake version 3.22.1
    • ninja version 1.10.1
    • pkg-config version 0.29.2

[✓] Android Studio (version 2021.1)
    • Android Studio at /home/huynq/Documents/android-studio
    • Flutter plugin version 67.0.1
    • Dart plugin version 211.7817
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)

[!] Android Studio (version unknown)
    • Android Studio at /snap/android-studio/current
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • android-studio-dir = /snap/android-studio/current
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] VS Code (version 1.84.2)
    • VS Code at /usr/share/code
    • Flutter extension version 3.90.0

[✓] VS Code (version 1.90.1)
    • VS Code at /snap/code/current/usr/share/code
    • Flutter extension version 3.90.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 22.04.3 LTS 6.5.0-35-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 120.0.6099.199

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
```
</details>

**Comment by justinmc** [comment]
Do we know what the native behavior is for sure on each of our platforms?

@KaushikGupta007 Just as a temporary workaround, would it help if you made the readOnly field non-focusable?

**Comment by KaushikGupta007** [comment]
@justinmc i cannot make `readOnly` field non-focusable as it will not solve purpose. Only temporary workaround i can think of is wrapping my `readOnly` field in `Shortcuts` and define `Enter Action` to replicate `onFieldSubmitted` functionality. But It would be better if this issue is fixed.

**Comment by tgrushka** [comment]
Hi guys, I think I found the problem!

TLDR: `bool get _shouldCreateInputConnection => kIsWeb || !widget.readOnly;` in 
`editable_text.dart`. I changed this to always be `true` and it fixed both:
- enables Edit -> Copy and Edit -> Select All (but also enables Cut, Paste, etc. so just setting it to `true` is not the solution);
- triggers `onFieldSubmitted` when I press Enter/Return on the `readOnly` field.

So it appears `TextInputConnection` is critical to these types of functions.

There's an **accessibility** issue that seems related to this one: on macOS desktop, the Edit menu bar -> **Copy** and **Select All** entries are not enabled when a `readOnly` text input is focused / selected. This is a problem because my custom screen readers send a ⌘-C (macOS) or `xsel` (Linux -- haven't tested yet) to get the current selected text and read it aloud. (My only workaround would be to enclose each field in ⌘-C shortcut.) Nonetheless, it's bad, because the menu bar behavior is inconsistent / unavailable only because of `readOnly`.

I drilled down to `TextEditable` and found that, when `readOnly` is true, a `TextInputConnection` is not created on desktop.

Here's relevant code in `editable_text.dart` (line ~2307):
```dart
  /// Whether to create an input connection with the platform for text editing
  /// or not.
  ///
  /// Read-only input fields do not need a connection with the platform since
  /// there's no need for text editing capabilities (e.g. virtual keyboard).
  ///
  /// On the web, we always need a connection because we want some browser
  /// functionalities to continue to work on read-only input fields like:
  ///
  /// - Relevant context menu.
  /// - cmd/ctrl+c shortcut to copy.
  /// - cmd/ctrl+a to select all.
  /// - Changing the selection using a physical keyboard.
  bool get _shouldCreateInputConnection => kIsWeb || !widget.readOnly;
```

**Comment by KaushikGupta007** [comment]
Hi team,

I wanted to follow up on this issue to see if there has been any progress or if additional information is needed from my side to help with the resolution. I understand that you're likely busy with many other requests, but I'd appreciate any updates or guidance on next steps.

Thank you for your time and support!

**Comment by bleroux** [comment]
It might be tricky to change this behavior because `TextField.onSubmitted` is coupled to the editing lifecycle ([from the documentation](https://api.flutter.dev/flutter/material/TextField/onSubmitted.html):  "Called when the user indicates that they are done editing the text in the field.").

Maybe overriding the enter key shortcut could help for this use case:

<details><summary>Updated code sample with custom shortcut for enter key</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

Future<void> main() async {
  runApp(MaterialApp(
    initialRoute: "/",
    routes: {
      "/": (context) => const Home(),
    },
  ));
}

class ValidateIntent extends Intent {
  const ValidateIntent();
}

class Home extends StatefulWidget {
  const Home({super.key});

  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Flutter v3.22.2"),
      ),
      body: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Column(
          children: [
            Shortcuts(
              shortcuts: const <ShortcutActivator, Intent>{
                SingleActivator(LogicalKeyboardKey.enter): ValidateIntent(),
              },
              child: Actions(
                actions: <Type, Action<Intent>>{
                  ValidateIntent: CallbackAction<ValidateIntent>(
                    onInvoke: (ValidateIntent intent) {
                      var snackBar = const SnackBar(content: Text('Hello World'));
                      ScaffoldMessenger.of(context).showSnackBar(snackBar);
                      FocusScope.of(context).nextFocus();
                      return null;
                    },
                  ),
                },
                child: TextFormField(
                  decoration: const InputDecoration(hintText: "First Field", labelText: "First Field"),
                  readOnly: true,
                ),
              ),
            ),
            TextFormField(
              decoration: const InputDecoration(hintText: "Second Field", labelText: "Second Field"),
              readOnly: true,
            ),
          ],
        ),
      ),
    );
  }
}

```

</details>

**Comment by KaushikGupta007** [comment]
But it works as expected on web. It think should remain consistent for all the platforms.
